### PR TITLE
Add RecipePage model including more advanced StreamField & forms usage

### LINKED ISSUE #168: More complex StreamField instances
We deliberately aimed to keep the Bakery demo as simple as possible with the implementation of StreamField, but that has an unintended consequence that now the demo is used as part of the vagrant-wagtail-dev environment it's easy to miss things that only happen when StreamField elements have been nested/ use lots of different field types etc.

Currently my work-around has been to use a pre-existing StreamField from another project to introduce the complexity but that's sub-optimal as it means sharing code snippets around when working with a group of people (so that we're looking at a consistent implementation)

Perhaps we could introduce a new content type ('developer test page') with a more complex implementation of it?

### PR DESCRIPTION
**Pull request by thibaudcolas**
Fixes #168. Covers the same scenarios as #368 minus "StreamField in InlinePanel", with a particular focus on making the data model and content realistic enough to be reusable for all types of demos (including with clients).

This introduces a `recipes` app with a new Recipe page type that’s very similar to BlogPage – but with more advanced patterns. I chose to retain almost all of what BlogPage does currently and add on top, so we can demo both "simple" StreamField and more complex patterns all on the one page. The RecipePage also has very simple rendering of its contents in place, again so it’s more usable in demos.

I initially also wanted to include demos of Documents in there but decided against it as it felt too complex for the kind of time I had allocated for this.

---

Content I added:

- https://en.wikibooks.org/wiki/Cookbook:Mincemeat_Tart
- https://en.wikibooks.org/wiki/Cookbook:Hot_Cross_Bun
- https://en.wikibooks.org/wiki/Cookbook:Southern_Cornbread

### CODE REVIEW

File: bakerydemo/base/fixtures/bakerydemo.json [hunk 1]
```diff
@@ -2101,7 +2128,7 @@
     "fields": {
       "path": "00010002",
       "depth": 2,
-      "numchild": 6,
+      "numchild": 7,
```
Comment (thibaudcolas) [inline bakerydemo/base/fixtures/bakerydemo.json:2131]: The RecipeIndexPage sits under the homepage, and I’ve reordered those child pages so it’s in a more natural place in the site’s main nav.

File: bakerydemo/base/fixtures/bakerydemo.json [hunk 2]
```diff
@@ -2707,6 +2734,142 @@
...
+      "path": "000100020005",
+      "depth": 3,
+      "numchild": 3,
+      "translation_key": "fa20c55f-7ea0-45a6-8d45-67695ea1ef81",
+      "locale": 1,
+      "latest_revision": null,
+      "live": true,
+      "has_unpublished_changes": false,
+      "first_published_at": "2019-02-10T16:25:47.179Z",
+      "last_published_at": null,
```
Comment (thibaudcolas) [inline bakerydemo/base/fixtures/bakerydemo.json:2750]: I have backdated those pages and removed their revisions like we do for other pages.

File: bakerydemo/recipes/blocks.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+from bakerydemo.base.blocks import BlockQuote, HeadingBlock, ImageBlock
+
+
+class RecipeStepBlock(StructBlock):
+    text = RichTextBlock(features=["bold", "italic", "link"])
+    difficulty = ChoiceBlock(
+        widget=forms.RadioSelect,
+        choices=[("S", "Small"), ("M", "Medium"), ("L", "Large")],
+        default="S",
+    )
```
Comment (thibaudcolas) [inline bakerydemo/recipes/blocks.py:28]: This is a bit of a stretch in terms of realism but it’s important we have at least one StructBlock to demo.

File: bakerydemo/recipes/blocks.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+    class Meta:
+        template = 'blocks/recipe_step_block.html'
+        icon = 'tick'
+
+
+class RecipeStreamBlock(StreamBlock):
```
Comment (thibaudcolas) [inline bakerydemo/recipes/blocks.py:35]: Additions compared to existing BaseStreamBlock:

- Groups
- TableBlock
- TypedTableBlock
- ListBlock
- `min_num` and `max_num` validation

File: bakerydemo/recipes/models.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    )
+    person = models.ForeignKey(
+        "base.Person",
+        related_name="person_recipe_relationship",
+        on_delete=models.CASCADE,
+    )
+    panels = [FieldPanel("person")]
+
+
+class RecipePage(Page):
```
Comment (thibaudcolas) [inline bakerydemo/recipes/models.py:36]: Additions compared to BlogPage:

- `block_counts` validation for specific block types
- `help_text` for StreamField
- Rich text field with limited set of formatting options.
- HelpPanel
- StreamField inside a MultiFieldPanel
- `title` FieldPanel outside StreamField

File: bakerydemo/recipes/models.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+    # Specifies parent to Recipe as being RecipeIndexPages
+    parent_page_types = ["RecipeIndexPage"]
+
+    # Specifies what content types can exist as children of RecipePage.
+    # Empty list means that no child content types are allowed.
+    subpage_types = []
+
+
+class RecipeIndexPage(Page):
```
Comment (thibaudcolas) [inline bakerydemo/recipes/models.py:123]: Copy of BlogIndexPage.

File: bakerydemo/recipes/models.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+from wagtail.models import Orderable, Page
+from wagtail.search import index
+
+from bakerydemo.base.blocks import BaseStreamBlock
+from bakerydemo.base.models import Person
+
+from .blocks import RecipeStreamBlock
+
+
+class RecipePersonRelationship(Orderable, models.Model):
```
Comment (thibaudcolas) [inline bakerydemo/recipes/models.py:14]: Copy of the authors setup for blog posts.

File: bakerydemo/static/css/main.css [hunk 8]
```diff
@@ -404,6 +404,17 @@ blockquote .attribute-name {
...
+/* stylelint-disable-next-line selector-class-pattern */
+:is(.block-table_block, .block-typed_table_block) :is(table, tr, td, th) {
+  border: 1px solid currentColor;
+  padding: 0.25em;
+}
+
+/* stylelint-disable-next-line selector-class-pattern */
+:is(.block-table_block, .block-typed_table_block) table {
+  margin-bottom: 30px;
+}
```
Comment (thibaudcolas) [inline bakerydemo/static/css/main.css:416]: Minimal styles so the tables look the part.

File: bakerydemo/templates/blocks/recipe_step_block.html [hunk 9]
```diff
@@ -0,0 +1 @@
+{{ self.text }}
```
Comment (thibaudcolas) [inline bakerydemo/templates/blocks/recipe_step_block.html:1]: This doesn’t display the difficulty, I didn’t think it was worth the trouble.

File: bakerydemo/templates/recipes/recipe_index_page.html [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+                    {% include "includes/card/blog-listing-card.html" with blog=recipe %}
+                {% endfor %}
+            {% else %}
+                <div class="col-md-12">
+                    <p>Oh, snap. Looks like we were too busy baking to write any recipes. Sorry.</p>
+                </div>
+            {% endif %}
+        </div>
+    </div>
+{% endblock content %}
```
Comment (thibaudcolas) [inline bakerydemo/templates/recipes/recipe_index_page.html:18]: Copy of the `blog_index_page.html`.

File: bakerydemo/templates/recipes/recipe_page.html [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+                {% endif %}
+
+                {# Give a label to the recipe section for screen reader users. #}
+                <div id="recipe-headline" class="sr-only">
+                    Recipe{% if page.recipe_headline %}: {{ page.recipe_headline|richtext }}{% endif %}
+                </div>
+
+                <section aria-labelledby="recipe-headline">
+                    {{ page.body }}
+                </section>
```
Comment (thibaudcolas) [inline bakerydemo/templates/recipes/recipe_page.html:36]: Aside from the backstory and headline, this is a copy of `blog_page.html`.

**Review by laymonage — APPROVED** [review]
Thanks, it's very useful to have these in place! I'm happy with this as-is, but I'm also up if you'd like to go through this together before merging.

**Pull request merged**